src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js

```diff
@@ -0,0 +1,100 @@
+ *  
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *  
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
+ * DEALINGS IN THE SOFTWARE.
+ * 
+ */
+
+/*eslint-env node */
+/*jslint node: true */
+
+"use strict";
+
+(function () {
+        
+    var fs = require("fs"),
+        _dirtyFilesCache = [];
+    
+    /**
+     * Clears the cache for dirty file paths
+     */
+    function clearDirtyFilesCache() {
+        _dirtyFilesCache = [];
+    }
+    
+    /**
+     * Updates the files cache with fullpath when dirty flag changes for a document
+     * If the doc is being marked as dirty then an entry is created in the cache
+     * If the doc is being marked as clean then the corresponsing entry gets cleared from cache
+     *
+     * @param {String} name - fullpath of the document
+     * @param {boolean} action - whether the document is dirty
+     */
+    function updateDirtyFilesCache(name, action) {
+        if (action) {
+            _dirtyFilesCache.push(name);
+        } else {
+            var index = _dirtyFilesCache.indexOf(name);
+            if (index >= 0) {
+                _dirtyFilesCache.splice(index, 1);
+            }
+        }
+    }
+    
+    /**
+     * Extract content locally from the file system used fs.readFile()
+     *
+     * @param {String} fileName - fullpath of the document
+     * @param {Function} callback - callback handle to post the content back
+     */
+    function _readFile(fileName, callback) {
+        fs.readFile(fileName, 'utf8', function (err, data) {
+            var content = "";
+            if (!err) {
+                content = data;
+            }
+            console.log("File read ", fileName);
+            callback.apply(null, [fileName, content]);
+        });
+    }
+    
+    /**
+     * Extracts file content for the given file name(1st param) and invokes the callback handle(2nd param) with 
+     * extracted file content. Content can be extracted locally from the file system used fs.readFile()
+     * or conditionally from main context(brackets main thread) by using the 3rd param 
+     *
+     * @param {String} fileName - fullpath of the document
+     * @param {Function} callback - callback handle to post the content back
+     * @param {Object} extractFromMainContext - content request handle wrapper from main thread
+     */
+    function extractContent(fileName, callback, extractFromMainContext) {
+        if (_dirtyFilesCache.indexOf(fileName) !== -1) {
```

Comment:
this should be an object `{}` and used like `_dirtyFilesCache[fileName]` instead of searching an array